internal/querycoordv2/meta/replica_manager_helper.go

```diff
@@ -0,0 +1,272 @@
+package meta
+
+import (
+	"sort"
+
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+)
+
+// collectionAssignmentHelper is a helper to manage the replica assignment in same collection.
+type collectionAssignmentHelper struct {
+	collectionID            typeutil.UniqueID
+	resourceGroupToReplicas map[string]*replicasInSameRGAssignmentHelper
+}
+
+// newCollectionAssignmentHelper creates a new collectionAssignmentHelper.
+func newCollectionAssignmentHelper(
+	collectionID typeutil.UniqueID,
+	rgToReplicas map[string][]*Replica,
+	rgs map[string]typeutil.UniqueSet,
+) *collectionAssignmentHelper {
+	resourceGroupToReplicas := make(map[string]*replicasInSameRGAssignmentHelper)
+	for rgName, replicas := range rgToReplicas {
+		resourceGroupToReplicas[rgName] = newReplicaAssignmentHelper(rgName, replicas, rgs[rgName])
+	}
+
+	helper := &collectionAssignmentHelper{
+		collectionID:            collectionID,
+		resourceGroupToReplicas: resourceGroupToReplicas,
+	}
+	helper.updateIncomingNodesAndExpectedNode()
+	return helper
+}
+
+// updateIncomingNodesAndExpectedNode updates the incoming nodes for all resource groups.
+// An incoming node is a node that not used by current collection but in resource group.
+func (h *collectionAssignmentHelper) updateIncomingNodesAndExpectedNode() {
+	// incoming nodes should be compared with all node of replica in same collection, even not in same resource group.
+	for _, helper := range h.resourceGroupToReplicas {
+		// some node in current resource group may load other replica data of same collection in other resource group.
+		// those node cannot be used right now.
+		newIncomingNodes := helper.nodesInRG.Clone()
+		currentUsedNodeCount := newIncomingNodes.Len()
+		h.RangeOverReplicas(func(rgName string, assignment *replicaAssignmentInfo) {
+			assignment.RangeOverAllNodes(func(nodeID int64) {
+				if newIncomingNodes.Contain(nodeID) {
+					newIncomingNodes.Remove(nodeID)
+					if rgName != helper.rgName {
+						// Node is still used by other replica of same collection in other resource group, cannot be used right now.
+						// filter it out to calculate the expected node count to avoid node starve of some replica in same resource group.
+						currentUsedNodeCount--
+					}
+				}
+			})
+		})
+		helper.incomingNodes = newIncomingNodes
+		helper.updateExpectedNodeCountForReplicas(currentUsedNodeCount)
```

Comment:
add a log after updating the number of nodes in replica for each rg?

Reply:
will add new log at replica manager.